policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+            else:
+                # else define service policy
+                associations = ep.generate_svc_profiles_pods()
+                for profile, pods in associations.items():
+                    # find pods and append profiles
+                    for pod in pods:
+                        existing_pod = self.match_pod(
+                            namespace=ep.namespace, name=pod)
+                        _log.debug("Adding profile %s to pod %s" % (profile, existing_pod))
+                        if existing_pod:
+                            try:
+                                _datastore_client.append_profiles_to_endpoint(profile_names=[profile],
+                                                                              endpoint_id=existing_pod.ep_id)
+                            except ProfileAlreadyInEndpoint:
+                                _log.warning("Applying %s to Pod %s : Profile Already exists" % \
+                                             (profile, existing_pod.key))
+                        else:
+                            _log.warning("Pod %s is not yet processed" % (pod))
+
+                # declare Endpoints obj processed
+                self.epResourceVersion = ep.resourceVersion
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+
+
+class Resource():
+    """
+    Resource objects pull pertinent info from json blobs and maintain universal functions
+    """
+
+    def __init__(self, json):
+        """
+        On init, each Resource saves the raw json, pulls necessary info (unique),
+        and defines a unique key identifier
+        """
+        self._json = json
+        self.init_from_json(json)
+        self.resourceVersion = json["metadata"]["resourceVersion"]
+
+    def init_from_json(self, json):
+        """
+        Abstract Method.
+        Sets public variables relevant to the child Resource.
+        """
+        raise NotImplementedError("init_from_json not implemented.")
+
+    @property
+    def key(self):
+        """
+        Abstract method.
+        Determines a unique identifier for the Resource
+        """
+        raise NotImplementedError("@property key not implemented.")
+
+    def __str__(self):
+        return "%s: %s\n%s" % (self.kind, self.key, self._json)
+
+
+class Namespace(Resource):
+
+    def init_from_json(self, json):
+        self.kind = KIND_NAMESPACE
+        self.name = json["metadata"]["name"]
+        self.profile_name = "namespace_%s" % self.name
+
+        try:
+            self.policy = json["spec"]["experimentalNetworkPolicy"]
+        except KeyError:
+            _log.warning("Namespace does not have policy, assumed Open")
+            self.policy = "Open"
+
+    @property
+    def key(self):
+        return self.name
+
+    def create_ns_profile(self):
```

Comment:
`create_ns_profile` -> `create_profile`

The "ns" bit is implied, since we're operating on a namespace object.